### PR TITLE
changed cloning library to rfdc for runtime module

### PR DESCRIPTION
**Pull request by anshumanr**
- [x ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

## Proposed changes

switched cloning library from `github.com/pvorb/clone` to `github.com/davidmarkclements/rfdc` in runtime, to improve modify performance  for large number of nodes (in 1000s). Changes and flame-graphs [discussed in discource forum](https://discourse.nodered.org/t/cloning-performance-using-rfdc-instead-of-clone/81288/9)

## Checklist

- [x ] I have read the [contribution guidelines](https://github.com/node-red/node-red/blob/master/CONTRIBUTING.md)
- [ x] For non-bugfix PRs, I have discussed this change on the forum/slack team.
- [x ] I have run `grunt` to verify the unit tests pass
- [ ] I have added suitable unit tests to cover the new/changed functionality

### CODE REVIEW

File: package-lock.json
Comment (knolleary) [inline package-lock.json:1]: We don't include a package-lock.json file in the repo. Please remove from this PR.

**Comment by knolleary** [comment]
Thanks for this contribution.

Looking at the issue list of `rfdc` I found this issue which highlights a functional difference between `rfdc` and `clone` - https://github.com/davidmarkclements/rfdc/issues/29

In the places we're cloning JSON, that won't be a problem. But given you've done a global replace of clone, we need to carefully review each instance to make sure this difference in behaviour won't have any impact.

**Comment by anshumanr** [comment]
I have reduced the scope of the change to limit it to `\runtime\lib\flows\` only. From what I saw, it should be restricted to JSON payloads (but my knowledge is limited). It still has performance benefits. I attached more flame graphs to show the diff.

dev branch
nr-dev-fg

after change
nr-rfdc-fg

Let me know if there's something else I can do to make it easier.

**Comment by anshumanr** [comment]
Hi. I am still wondering if there's anything I can do to get this through. Besides the test cases in the code base, is there any other testing available which I can do to verify further?

**Comment by knolleary** [comment]
Hi @anshumanr 

It's just waiting on me right now to fully review and make sure there are no unexpected side effects. I'm clearing a backlog of issues and PRs this week - we'll get there.

**Comment by anshumanr** [comment]
Great. Thanks for the update. If possible, please let know if this PR is problematic.

**Comment by anshumanr** [comment]
Any update? as in -  "it was looked at but was problematic" or "yet to be looked at"

**Comment by knolleary** [comment]
@anshumanr sorry for the delay, this is definitely on the radar for 4.0. I just need to review the places you've swapped the modules and verify they are only being used to clone pure JSON. As long as that looks good, I see no reason not to get this merged.

**Comment by knolleary** [comment]
Hi @anshumanr 

Thanks for this - and appreciate your patience. We're getting Node-RED 4.0 ready at the moment and this looks good to me.

**Comment by knolleary** [comment]
Except... just as I click merge I realise something it missed... will fixup now

**Comment by knolleary** [comment]
Definitely merged this too quickly - failed to spot the tests hadn't run.

Since merging, tests now failing with:

```

  1) subflow
       should access typed value of env var:

      AssertionError: expected false == true
      + expected - actual

      -false
      +true

      at Node._inputCallback (test/nodes/subflow/subflow_spec.js:293:28)
      at /Users/nol/code/node-red/node-red/packages/node_modules/@node-red/runtime/lib/nodes/Node.js:214:26
      at Object.trigger (packages/node_modules/@node-red/util/lib/hooks.js:166:13)
      at Node._emitInput (packages/node_modules/@node-red/runtime/lib/nodes/Node.js:206:11)
      at Node.emit (packages/node_modules/@node-red/runtime/lib/nodes/Node.js:190:25)
      at Node.receive (packages/node_modules/@node-red/runtime/lib/nodes/Node.js:499:10)
      at deliverMessageToDestination (packages/node_modules/@node-red/runtime/lib/flows/Flow.js:800:40)
      at Immediate._onImmediate (packages/node_modules/@node-red/runtime/lib/flows/Flow.js:816:21)
      at process.processImmediate (node:internal/timers:476:21)
```

in `test/nodes/subflow/subflow_spec.js`

Going to revert this change and investigate.

**Comment by anshumanr** [comment]
ahh! did i miss something? I do remember running the unit test cases.

**Comment by anshumanr** [comment]
It seems some more test cases have been added since which fail. If i go to my branch and run the tests again, they continue to pass.

```
subflow
    ✔ should define subflow
    ✔ should pass data to/from subflow
    ✔ should pass data to/from nested subflow
    ✔ should access env var of subflow template
    ✔ should access env var of subflow instance
    ✔ should access last env var with same name
    ✔ should access typed value of env var
    ✔ should overwrite env var of subflow template by env var of subflow instance
    ✔ should access env var of parent subflow template
    ✔ should access env var of parent subflow instance
    ✔ should access env var of tab
    ✔ should access env var of group
    ✔ should access env var of nested group
    ✔ should access NR_NODE_PATH env var within subflow instance


  2451 passing (2m)
  59 pending


Done.

=============================== Coverage summary ===============================
Statements   : 68.69% ( 11546/16809 )
Branches     : 61.5% ( 6368/10354 )
Functions    : 64.89% ( 1619/2495 )
Lines        : 68.8% ( 11195/16271 )
================================================================================

Done.
```

I will update branch and test again. Will file another PR.

**Comment by knolleary** [comment]
I have merge the PR that reverts the entire set of changes to keep the dev branch clean. Please do send over a new PR that reapplies the changes plus whatever is needed to fix the tests.

**Comment by anshumanr** [comment]
https://github.com/node-red/node-red/pull/4607 opened for the same

**Pull request merged**